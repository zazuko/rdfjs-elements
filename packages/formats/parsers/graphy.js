async function parser(module) {
  const create = (await module).default

  return {
    import(quadStream, options) {
      return quadStream.pipe(create(options))
    },
  }
}

export const trig = () => parser(import('@graphy/content.trig.read'))
export const nq = () => parser(import('@graphy/content.nq.read'))
