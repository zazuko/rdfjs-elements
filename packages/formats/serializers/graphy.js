import { lazySink } from '@zazuko/formats-lazy/LazySink.js'
import { TransformToConciseHash } from './TransformToConciseHash.js'

async function serializer(importScribe, { strict, ...defaults } = {}) {
  const create = (await importScribe).default

  return class {
    // eslint-disable-next-line class-methods-use-this
    import(quadStream, { preserveListNodeProperties, ...options } = {}) {
      const prefixes = {
        ...(defaults.prefixes || {}),
        ...(options.prefixes || {}),
      }

      const writer = create({
        prefixes,
      })

      quadStream
        .pipe(
          new TransformToConciseHash({
            prefixes,
            strict,
            preserveListNodeProperties,
          })
        )
        .pipe(writer)

      return writer
    }
  }
}

export const TurtleSerializer = lazySink(() =>
  serializer(import('@graphy/content.ttl.write'))
)
export const RdfXmlSerializer = lazySink(() =>
  serializer(import('@graphy/content.xml.scribe'), { strict: true })
)
export const TrigSerializer = lazySink(() =>
  serializer(import('@graphy/content.trig.write'), { strict: true })
)
