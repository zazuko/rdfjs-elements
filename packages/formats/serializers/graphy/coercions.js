import { xsd } from '@tpluscode/rdf-ns-builders'
import { shrink } from '@zazuko/prefixes'

export class LongLiteral {
  constructor(term) {
    this.term = term
  }

  toTerm() {
    const raw = `"""${this.term.value.replace(/"$/, '\\"')}"""`

    return {
      terse: prefixes => raw + this.langOrDatatype(prefixes),
      verbose: prefixes => raw + this.langOrDatatype(prefixes),
    }
  }

  langOrDatatype(prefixes) {
    if (this.term.language) {
      return `@${this.term.language}`
    }

    if (this.term.datatype.equals(xsd.string)) {
      return ''
    }

    const shrunk = shrink(this.term.datatype.value, prefixes)
    if (!shrunk) {
      return `^^<${this.term.datatype.value}>`
    }

    return `^^${shrunk}`
  }
}

export const coercions = new Map([[LongLiteral, literal => literal.toTerm()]])
