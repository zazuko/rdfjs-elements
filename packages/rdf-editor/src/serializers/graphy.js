async function serializer(importScribe) {
  const create = (await importScribe).default

  return {
    import(quadStream, { prefixes = {} } = {}) {
      const writer = create({
        prefixes,
      })
      writer.import(quadStream)

      return writer
    },
  }
}

export const turtle = () => serializer(import('@graphy/content.ttl.scribe'))
export const rdfXml = () => serializer(import('@graphy/content.xml.scribe'))
export const trig = () => serializer(import('@graphy/content.trig.scribe'))
