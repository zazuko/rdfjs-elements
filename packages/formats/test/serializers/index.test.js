import { expect } from 'chai'
import prefixes from '@zazuko/prefixes'
import * as ns from '@tpluscode/rdf-ns-builders'
import $rdf from 'rdf-ext'
import clownface from 'clownface'
import getStream from 'get-stream'
import rdfUtil from 'rdf-utils-fs'
import { dirname, join } from 'path'
import { fileURLToPath } from 'url'
import JsonldSerializer from '../../serializers/jsonld.js'
import { TurtleSerializer } from '../../serializers/graphy.js'

const turtle = new TurtleSerializer()
const jsonld = new JsonldSerializer()

const __dirname = dirname(fileURLToPath(import.meta.url))

describe('@rdfjs-elements/formats-pretty/serializers', () => {
  describe('turtle', () => {
    it('combines default and import prefixes', async () => {
      // given
      const { dcterms, schema } = prefixes
      const sink = await turtle({
        prefixes: { schema },
      })
      const dataset = $rdf.dataset()
      clownface({ dataset })
        .namedNode('https://example.com/john')
        .addOut(ns.dcterms.type, ns.schema.Person)

      // when
      const serialized = await getStream(
        sink.import(dataset.toStream(), {
          prefixes: { dcterms },
        })
      )

      // then
      expect(serialized).to.contain(`prefix schema: <${schema}>`)
      expect(serialized).to.contain(`prefix dcterms: <${dcterms}>`)
      expect(serialized).to.contain(`dcterms:type schema:Person`)
    })

    it('removes excess rdf List node properties by default', async () => {
      // given
      const { owl, as } = prefixes
      const graph = rdfUtil.fromFile(
        join(__dirname, `../graphs/list-with-extras.ttl`)
      )
      const sink = await turtle({
        prefixes: { owl, as },
      })

      // when
      const serialized = await getStream(sink.import(graph))

      // then
      expect(serialized).to.match(/owl:unionOf \(\s+as:Object\s+as:Link\s+\)/m)
    })

    it('does not remove excess rdf List node properties when option set', async () => {
      // given
      const { owl, as, rdf, rdfs } = prefixes
      const graph = rdfUtil.fromFile(
        join(__dirname, `../graphs/list-with-extras.ttl`)
      )
      const sink = await turtle({
        prefixes: { owl, as, rdf, rdfs },
      })

      // when
      const serialized = await getStream(
        sink.import(graph, { preserveListNodeProperties: true })
      )

      // then
      expect(serialized).to.match(
        /rdf:first as:Object ;\s+rdf:type rdfs:Resource, rdf:List ;/m
      )
      expect(serialized).to.match(
        /rdf:first as:Link ;\s+rdf:type rdf:List ;\s+rdf:rest rdf:nil ;/m
      )
    })

    it('serializes successfully a complex graph with reused nodes', async () => {
      // given
      const { owl, as, rdf, rdfs } = prefixes
      const graph = rdfUtil.fromFile(join(__dirname, `../graphs/view.ttl`))
      const sink = await turtle({
        prefixes: { owl, as, rdf, rdfs },
      })

      // when
      const serialized = await getStream(sink.import(graph))

      // then
      expect(serialized).to.be.ok
    })
  })

  describe('jsonld', () => {
    it('combines default and import prefixes', async () => {
      // given
      const { dcterms, schema } = prefixes
      const sink = await jsonld({
        prefixes: { schema },
      })
      const dataset = $rdf.dataset()
      clownface({ dataset })
        .namedNode('https://example.com/john')
        .addOut(ns.dcterms.type, ns.schema.Person)

      // when
      const serialized = await getStream(
        sink.import(dataset.toStream(), {
          prefixes: { dcterms },
        })
      )

      // then
      const json = JSON.parse(serialized)
      expect(json).to.deep.include({
        '@context': {
          schema,
          dcterms,
        },
      })
      expect(json).to.have.deep.property('dcterms:type', {
        '@id': 'schema:Person',
      })
    })
  })
})
