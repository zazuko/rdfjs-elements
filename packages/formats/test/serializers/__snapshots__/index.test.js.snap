// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@rdfjs-elements/formats-pretty/serializers turtle preserves line breaks in literals 1`] = `
"
<http://example.org/a> <http://example.org/b> \\"\\"\\"The first line
The second line
  more\\"\\"\\" .

<http://example.org/x> <http://example.org/y> \\"\\"\\"Long
description
with datatype\\"\\"\\"^^<http://example.org/multiline> .

<http://example.org/m> <http://example.org/n> \\"\\"\\"multiline
with language\\"\\"\\"@en-US .

<http://example.org/f> <http://example.org/g> \\"\\"\\"The first line
\\"The quoted line\\"
  more\\"\\"\\" .

<http://example.org/r> <http://example.org/q> \\"\\"\\"Quote
at end\\\\\\"\\"\\"\\" .

"
`;

exports[`@rdfjs-elements/formats-pretty/serializers turtle preserves line breaks in literals when using prefixes 1`] = `
"@prefix ex: <http://example.org/> .

ex:a ex:b \\"\\"\\"The first line
The second line
  more\\"\\"\\" .

ex:x ex:y \\"\\"\\"Long
description
with datatype\\"\\"\\"^^ex:multiline .

ex:m ex:n \\"\\"\\"multiline
with language\\"\\"\\"@en-US .

ex:f ex:g \\"\\"\\"The first line
\\"The quoted line\\"
  more\\"\\"\\" .

ex:r ex:q \\"\\"\\"Quote
at end\\\\\\"\\"\\"\\" .

"
`;
